### PR TITLE
Verify Tree-SHA512s in merge commits, enforce sigs are not SHA1

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This will validate Tree-SHA512s in merge commits if verify-commits.sh is called with --tree-checks as a second arg (eg verify-commits.sh upstream/master --tree-checks), as added by #9871. Note that I think we need to fix handling of symlinks in #9871 first.

Additionally, it disables sha1 for signature checking.

Due to some SHA1s in folks' selfsigs, the second part currently fails (@MarcoFalke, @sipa, @jonasschnelli.....)

### CODE REVIEW

**Comment by petertodd** [comment]
While concept ACK, note that collision attacks on sha1 signatures - particularly in selfsigs - aren't as insecure as you might think. The reason is they're basically an attack on yourself. For instance, in the case of a selfsig, you already can add multiple subkeys to your PGP identity anyway, and you can do so without revealing that fact to the public, so there isn't a direct attack vector.

Similarly, given that a git commit signature is an expression of trust anyway, creating two commit objects so you can sign both with the same OpenPGP signature isn't a clear attack on anything; unlike the case where someone inserts a malicious *file* into the Git tree, that can only be done by the signer

Meanwhile, with second-preimage attacks that third-parties can do, [almost every cryptographic hash function in history has proven immune](https://z.cash/technology/history-of-hash-function-attacks.html). In fact, in my OpenTimestamps Python library I note that [even MD5 is secure enough for timestamping](https://github.com/opentimestamps/python-opentimestamps/blob/b395fcc25354184cb2c6a3a899f671f62af4ab44/opentimestamps/core/op.py#L311) (though I'm still not inclined to add support for it out of a healthy sense of caution!).

Now there *are* more complex cases where this stuff matters - timestamping is one of them depending on how it's implemented - so we probably should still merge this for the long-term. But we can certainly wait until people fix their keys.

**Comment by petertodd** [comment]
Oh, and another reason to Concept ACK this: it wouldn't be surprising if a colliding self-sig or similar could be used to trigger a non-cryptographic bug in validation somewhere, given that you're violating the common assumption that hash functions don't collide.

**Comment by TheBlueMatt** [comment]
@petertodd indeed, selfsig attacks aren't so interesting, but I'm not aware of another way to prevent sha1-based signatures. While probably also not particularly interesting attack vectors, its good belt-and-suspenders when signing "untrusted input" (in the form of the git tree root/sha512 tree root/commit messages/etc) to avoid sha1 inputs.

**Comment by petertodd** [comment]
@TheBlueMatt Oh, yeah you make a good point in saying this is a case of "untrusted input", as the commit message to be made is semi-deterministic and under the control of the pull-requester.

One sec, I'll write up a comment for you. :)

**Comment by TheBlueMatt** [comment]
* Rebased.
* Now enforces lack of symlinks in github-merge.py.
* Now enforces lack of symlinks in verify-commits.sh
* Always enforce non-SHA1-sigs even when --tree-checks is off
* Update trusted-sha256-root-commit until Marco updates his subkeys.

**Comment by laanwj** [comment]
Concept ACK

**Comment by laanwj** [comment]
Note that verify-commits on master is currently failing. I've tried it with this version and it doesn't pass either:
```
Using verify-commits data from /.../bitcoin/contrib/verify-commits.new
No parent of 58861ad91b499f6902c0f1c51bb745b91a275826 was signed with a trusted key!
Parents are:
commit f7ec7cfd38b543ba81ac7bed5b77f9a19739460b
Merge: 65d90f5 7ed143c
Author: MarcoFalke <falke.marco@gmail.com>
Date:   Thu Mar 2 22:30:12 2017 +0100

    Merge #9359: Add test for CWalletTx::GetImmatureCredit() returning stale values.
    
    7ed143c Add test for CWalletTx::GetImmatureCredit() returning stale values. (Russell Yanofsky)
    
    Tree-SHA512: c95088ed6dfc5a0774ddaa2fe14ac0a9ebd830922a4d77100ec3d51fdeb6df40ad97de4f2ea970ed0f4122dcc0022ee1d43ab3c7188becd7f90c1c6af0ed39b7
commit 64854666f57d9f94269000f10897858e8c60d99f
Author: Wladimir J. van der Laan <laanwj@gmail.com>
Date:   Thu Mar 2 14:49:11 2017 +0100

    test: Report InitBlockIndex result
    
    If InitBlockIndex fails, then it will segfault later. Same for the later
    ActivateBestChain. BOOST_REQUIRE the result, so that an error will be
    reported and the test case aborted.
```

**Comment by TheBlueMatt** [comment]
* Updated the trusted-sha256-root-commit for Jonas' sha2 commit.
* Allow any subkey that is properly signed, instead of only listing specific subkeys (this should fix verify-commits on master).

File: contrib/verify-commits/verify-commits.sh [hunk 1]
```diff
@@ -9,26 +9,84 @@
...
 	if [ $1 = $VERIFIED_ROOT ]; then
 		return 0;
 	fi
+
+	VERIFY_TREE=$2
+	NO_SHA1=$3
+	if [ $1 = $VERIFIED_SHA512_ROOT ]; then
+		VERIFY_TREE=0
+		NO_SHA1=0
+		if [ "$VERIFY_TREE" = "1" ]; then
```
Comment (MarcoFalke) [inline contrib/verify-commits/verify-commits.sh]: Either the constant is off by one or the line this is placed in is off by two.

**Comment by MarcoFalke** [comment]
utACK 6b2a4a3

File: contrib/verify-commits/verify-commits.sh (hunk 1 above)
Comment (TheBlueMatt) [inline contrib/verify-commits/verify-commits.sh]: Oops, indeed. Fixed and squashed.

**Comment by TheBlueMatt** [comment]
We may want to backport at least the "Allow any subkey in verify-commits" commit for 0.14 so that verify-commits passes on release

**Comment by TheBlueMatt** [comment]
Pushed a new commit which will fix the checks on the 0.14 branch.

**Comment by jonasschnelli** [comment]
Verified the new/changed keys (same as in #9920).
Tested ACK bbd757940bcb0628df6f7a5bd1fb348cf2290502

File: contrib/verify-commits/verify-commits.sh [hunk 2]
```diff
@@ -9,26 +9,84 @@
...
+					IFS="$IFS_CACHE"
+					return 1
+					;;
+			esac
+		done
+		IFS="$IFS_CACHE"
+
+		FILE_HASHES=""
+		for FILE in $(git ls-tree --full-tree -r --name-only $1 | LANG=C sort); do
+			HASH=$(git cat-file blob $1:"$FILE" | sha512sum | { read FIRST OTHER; echo $FIRST; } )
```
Comment (jonasschnelli) [inline contrib/verify-commits/verify-commits.sh]: nit: IMO `sha512sum` is not portable (at least it's not available on OSX). What about `shasum -a 512` instead?

**Comment by laanwj** [comment]
Going to merge this to fix travis. The shasum nit can be fixed later.

File: contrib/verify-commits/verify-commits.sh (hunk 2 above)
Comment (TheBlueMatt) [inline contrib/verify-commits/verify-commits.sh]: shasum is part of perl, so is probably less standard than sha512sum. We should have a check for whether sha512sum is available and only use shasum if it is not.

**Pull request merged**